packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts

```diff
@@ -113,35 +113,53 @@ export class PersistenceUserManager {
       );
     }
 
-    // Use the first persistence that supports a full read-write roundtrip (or fallback to memory).
-    let chosenPersistence = _getInstance<PersistenceInternal>(
-      inMemoryPersistence
-    );
-    for (const persistence of persistenceHierarchy) {
-      if (await persistence._isAvailable()) {
-        chosenPersistence = persistence;
-        break;
-      }
-    }
+    // Eliminate any persistences that are not available
+    const persistences = (
+      await Promise.all(
+        persistenceHierarchy.map(async persistence => {
+          if (await persistence._isAvailable()) {
+            return persistence;
+          }
+          return undefined;
+        })
+      )
+    ).filter(persistence => persistence) as PersistenceInternal[];
+
+    // Fall back to the first persistence listed, or in memory if none available
+    let chosenPersistence =
+      persistences[0] || _getInstance<PersistenceInternal>(inMemoryPersistence);
 
-    // However, attempt to migrate users stored in other persistences (in the hierarchy order).
-    let userToMigrate: UserInternal | null = null;
     const key = _persistenceKeyName(userKey, auth.config.apiKey, auth.name);
+
+    // Pull out the existing user, setting the chosen persistence to that
+    // persistence if the user exists.
+    let userToMigrate: UserInternal | null = null;
+    // Note, here we check for a user in _all_ persistences, not just the
+    // ones deemed available. If we can migrate a user out of a broken
+    // persistence, we will (but only if that persistence supports migration).
     for (const persistence of persistenceHierarchy) {
-      // We attempt to call _get without checking _isAvailable since here we don't care if the full
-      // round-trip (read+write) is supported. We'll take the first one that we can read or give up.
       try {
-        const blob = await persistence._get<PersistedBlob>(key); // throws if unsupported
+        const blob = await persistence._get<PersistedBlob>(key);
         if (blob) {
           const user = UserImpl._fromJSON(auth, blob); // throws for unparsable blob (wrong format)
-          if (persistence !== chosenPersistence) {
-            userToMigrate = user;
-          }
+          chosenPersistence = persistence;
```

Comment:
I'd suggest not assigning to `chosenPersistence` here since it is not the chosen persistence (but rather the persistence to migrate from). It is very confusing that `chosenPersistence` is then overwritten in L162 below. I'd suggest inlining `_shouldAllowMigration` check from below. Or just use a different variable name if you must.

Reply:
Renamed to "selectedPersistence" as that accurately reflects what this code is doing